compiler_opt/distributed/local/local_worker_manager.py

```diff
@@ -0,0 +1,220 @@
+              worker_class: 'type[Worker]', *args, **kwargs):
+  """Worker process entrypoint."""
+  pool = concurrent.futures.ThreadPoolExecutor()
+  obj = worker_class(*args, **kwargs)
+
+  def make_ondone(msgid):
+
+    def on_done(f: concurrent.futures.Future):
+      if f.exception():
+        out_q.put(TaskResult(msgid=msgid, success=False, value=f.exception()))
+      else:
+        out_q.put(TaskResult(msgid=msgid, success=True, value=f.result()))
+
+    return on_done
+
+  # Run forever. The stub will just kill the runner when done.
+  while True:
+    task = in_q.get()
+    the_func = getattr(obj, task.func_name)
+    application = functools.partial(the_func, *task.args, **task.kwargs)
+    if task.is_urgent:
+      try:
+        res = application()
+        out_q.put(TaskResult(msgid=task.msgid, success=True, value=res))
+      except BaseException as e:  # pylint: disable=broad-except
+        out_q.put(TaskResult(msgid=task.msgid, success=False, value=e))
+    else:
+      pool.submit(application).add_done_callback(make_ondone(task.msgid))
+
+
+def _run(*args, **kwargs):
+  try:
+    _run_impl(*args, **kwargs)
+  except BaseException as e:
+    logging.error(e)
+    raise e
+
+
+def _make_stub(cls: 'type[Worker]', *args, **kwargs):
+
+  class _Stub():
+    """Client stub to a worker hosted by a process."""
+
+    def __init__(self):
+      self._send: 'queue.Queue[Task]' = multiprocessing.get_context().Queue()
+      self._receive: 'queue.Queue[TaskResult]' = multiprocessing.get_context(
+      ).Queue()
+
+      # this is the process hosting one worker instance.
+      self._process = multiprocessing.Process(
+          target=functools.partial(
+              _run,
+              worker_class=cls,
+              in_q=self._send,
+              out_q=self._receive,
+              *args,
+              **kwargs))
+      # lock for the msgid -> reply future map
+      self._lock = threading.Lock()
+      self._map: Dict[int, concurrent.futures.Future] = {}
+      # thread drainig the receive queue
+      self._pump = threading.Thread(target=self._msg_pump)
+
+      # event set by _pump when it exits
+      self._done = threading.Event()
+      # event set by the stub to tell the message pump to exit.
+      self._stop_pump = threading.Event()
+
+      # atomic control to _msgid
+      self._msgidlock = threading.Lock()
+      self._msgid = 0
+
+      # start the worker and the message pump
+      self._process.start()
+      self._pump.start()
+
+    def _is_cancelled_or_broken(self):
+      return self._stop_pump.is_set() or not self._process.is_alive()
+
+    def _msg_pump(self):
+      while not self._is_cancelled_or_broken():
+        try:
+          # avoid blocking so we may notice if _stop_pump was set. We aren't
+          # very concerned with delay between shutdown request and the message
+          # pump noticing it, this happens only at the very end of training.
+          task_result = self._receive.get(timeout=1.0)
```

Comment:
Could avoid the timeout/try/except by feeding a poison pill (e.g. `None`) into the queue and breaking out of the loop if it's detected. Only edge case I can think of is externally killing the child process, not sure if you can feed a poison pill after that

Reply:
If the external process dies unintentionally (i.e. you got a bug - like illustrated in the test), then the poison pill won't ever come, and the pump will hang out forever.

Reply:
adding a `finally:` block which feeds the poison pill to `_run` should do the trick for that

Reply:
The main shortcoming would be the 1s timeout, which on average would be a 0.5s wait, and I believe we’ll be killing the processes fairly frequently too. On a 10s training cycle that’d be 5%

I think we’d only need to guarantee the correctness of _run outside the try/catch/finally block, which is trivially correct since it’s nothing. (I suppose starting the processes needs to be checked too for completeness) 
_run_impl is free to throw exceptions, since _run will put the poison pill in regardless because of the finally block (except, possibly, that one edge case I noted above)


Also, catching exceptions is [“slow”](https://stackoverflow.com/questions/2522005/cost-of-exception-handlers-in-python)

Reply:
We're not killing these processes ever except for at the very end of training.

Reply:
...but - we can just inject `None` at shutdown instead. Doing that.